pkg/virt-handler/node-labeller/kvm-caps-info-plugin.go

```diff
@@ -0,0 +1,247 @@
+package nodelabeller
+
+/*
+#include <linux/kvm.h>
+const int IoctlGetMsrIndexList = KVM_GET_MSR_INDEX_LIST;
+const int IoctlCheckExtension = KVM_CHECK_EXTENSION;
+// Capabilities (extensions).
+const int CapHyperv = KVM_CAP_HYPERV;
+const int CapHypervTime = KVM_CAP_HYPERV_TIME;
+const int CapHypervVpIndex = KVM_CAP_HYPERV_VP_INDEX;
+const int CapHypervTlbflush = KVM_CAP_HYPERV_TLBFLUSH;
+const int CapHypervSendIPI = KVM_CAP_HYPERV_SEND_IPI;
+const int CapHypervSynic = KVM_CAP_HYPERV_SYNIC;
+const int CapHypervSynic2 = KVM_CAP_HYPERV_SYNIC2;
+int msr_list_index(void *data, int n, __u32 *index) {
+    struct kvm_msr_list *msrs = (struct kvm_msr_list*)data;
+    if (n >= msrs->nmsrs) {
+        return 1;
+    }
+    *index = msrs->indices[n];
+    return 0;
+}
+*/
+import "C"
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+	"unsafe"
+	//gomsr "github.com/fearful-symmetry/gomsr"
+)
+
+const (
+	HV_X64_MSR_CRASH_CTL               = 0x40000105
+	HV_X64_MSR_RESET                   = 0x40000003
+	HV_X64_MSR_VP_INDEX                = 0x40000002
+	HV_X64_MSR_VP_RUNTIME              = 0x40000010
+	HV_X64_MSR_SCONTROL                = 0x40000080
+	HV_X64_MSR_STIMER0_CONFIG          = 0x400000B0
+	HV_X64_MSR_TSC_FREQUENCY           = 0x40000022
+	HV_X64_MSR_REENLIGHTENMENT_CONTROL = 0x40000106
+)
+
+type tmsr struct {
+	nmsrs   C.__u32
+	indices []C.__u32
+}
+
+type Msr struct {
+	Index uint32 `json:"index"`
+	Value uint64 `json:"value"`
+}
+
+type CapScan struct {
+	supprtedMSRS map[uint32]bool
+	fd           uintptr
+}
+
+type capability struct {
+	Name      string
+	Extension uintptr
+	MSR       int
+}
+
+func NewCapScanner() *CapScan {
+	return new(CapScan)
+}
+
+var CapsDesc = []capability{
+	{
+		Extension: uintptr(C.CapHyperv),
+		Name:      "base",
+	},
+	{
+		Extension: uintptr(C.CapHypervTime),
+		Name:      "time",
+	},
+	{
+		Extension: uintptr(C.CapHypervVpIndex),
+		MSR:       HV_X64_MSR_VP_INDEX,
+		Name:      "vpindex",
+	},
+	{
+		Extension: uintptr(C.CapHypervTlbflush),
+		Name:      "tlbflush",
+	},
+	{
+		Extension: uintptr(C.CapHypervSendIPI),
+		Name:      "ipi",
+	},
+	{
+		Extension: uintptr(C.CapHypervSynic),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic",
+	},
+	{
+		Extension: uintptr(C.CapHypervSynic2),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic2",
+	},
+	{
+		MSR:  HV_X64_MSR_TSC_FREQUENCY,
+		Name: "frequencies",
+	},
+	{
+		MSR:  HV_X64_MSR_RESET,
+		Name: "reset",
+	},
+	{
+		MSR:  HV_X64_MSR_VP_RUNTIME,
+		Name: "runtime",
+	},
+	{
+		MSR:  HV_X64_MSR_STIMER0_CONFIG,
+		Name: "synictimer",
+	},
+	{
+		MSR:  HV_X64_MSR_REENLIGHTENMENT_CONTROL,
+		Name: "reenlightenment",
+	},
+}
+
+func availableMsrs(fd uintptr) ([]uint32, error) {
```

Comment:
is it reasonably possible to unit test this? i guess we'd need a mock /dev/kvm device or something. maybe that's impractical though.

Reply:
How can we mock kvm device?